openedx/features/subscriptions/utils.py

```diff
@@ -23,16 +27,24 @@ def track_subscription_enrollment(subscription_id, user, course_id, site):
             valid_user_subscription.course_enrollments.add(enrollment)
             valid_user_subscription.save()
 
+
 def is_course_accessible_with_subscription(user, course):
     """
     Check if user has access to a course enrolled through subscription.
     """
+    if not waffle.switch_is_active(settings.ENABLE_SUBSCRIPTIONS_ON_RUNTIME_SWITCH):
+        return ACCESS_GRANTED
+
+    if not user or not user.is_authenticated:
+        return ACCESS_GRANTED
+
     course_enrolled_subscriptions = UserSubscription.objects.filter(user=user, course_enrollments__course__id=course.id)
     if not course_enrolled_subscriptions:
         return ACCESS_GRANTED
+    else:
+        for subscription in course_enrolled_subscriptions:
+            if subscription.is_active:
+                return ACCESS_GRANTED
 
-    for subscription in course_enrolled_subscriptions:
-        if subscription.is_active:
-            return ACCESS_GRANTED
+        return ACCESS_DENIED
```

Comment:
The condition would reach this location only if the given user had enrolled in a course using a subscription that is no longer active.